### PR TITLE
config: add path_config_source and watched_directory config

### PR DESCRIPTION
**Pull request by mattklein123**
For xDS over the file system, sometimes more control is required over
what directory/file is watched for symbolic link swaps. Specifically,
in order to deliver xDS over a Kubernetes ConfigMap, this extra
configuration is required.

Fixes https://github.com/envoyproxy/envoy/issues/10979

Risk Level: Low, new feature
Testing: Integration test
Docs Changes: Added
Release Notes: Added
Platform Specific Features: N/A
Deprecated: Yes, details added

### CODE REVIEW

File: api/envoy/config/core/v3/config_source.proto [hunk 1]
```diff
@@ -143,13 +144,48 @@ message RateLimitSettings {
...
+  //   demonstrated in the :ref:`runtime documentation <config_runtime_symbolic_link_swap>` can be
+  //   used here also. If `watched_directory` is configured, no watch will be placed directly on
+  //   this path. Instead, the configured `watched_directory` will be used to trigger reloads of
+  //   this path. This is required in certain deployment scenarios. See below for more information.
+  string path = 1 [(validate.rules).string = {min_len: 1}];
+
+  // If configured, this directory will be watched for *moves.* When an entry in this directory is
+  // moved to, the `path` will be reloaded. This is required in certain deployment scenarios.
+  //
+  // Specifically, if trying to load an xDS resource using a Kubernetes ConfigMap, the following
```
Comment (htuch) [inline api/envoy/config/core/v3/config_source.proto:170]: Can you link to a reference on ConfigMap?

File: envoy/router/router.h [hunk 2]
```diff
@@ -383,7 +383,7 @@ class RetryState {
   /**
    * Determine whether a request should be retried based on the response headers.
    * @param response_headers supplies the response headers.
-   * @param original_request supplies the orignal request headers.
+   * @param original_request supplies the original request headers.
```
Comment (htuch) [inline envoy/router/router.h:386]: Nit: unrelated, but OK.

File: source/common/config/filesystem_subscription_impl.cc [hunk 3]
```diff
@@ -12,18 +12,35 @@
...
-  });
+    Event::Dispatcher& dispatcher,
+    const envoy::config::core::v3::PathConfigSource& path_config_source,
+    SubscriptionCallbacks& callbacks, OpaqueResourceDecoder& resource_decoder,
+    SubscriptionStats stats, ProtobufMessage::ValidationVisitor& validation_visitor, Api::Api& api)
+    : path_(path_config_source.path()), callbacks_(callbacks), resource_decoder_(resource_decoder),
+      stats_(stats), api_(api), validation_visitor_(validation_visitor) {
+  if (!path_config_source.has_watched_directory()) {
+    file_watcher_ = dispatcher.createFilesystemWatcher();
+    file_watcher_->addWatch(path_, Filesystem::Watcher::Events::MovedTo, [this](uint32_t) {
```
Comment (htuch) [inline source/common/config/filesystem_subscription_impl.cc:30]: Can you map the legacy case to `WatchedDirectory` to dedupe this logic?

**Review by htuch — COMMENTED** [review]
API/implementation LGTM modulo minor comments, will look at tests tomorow.

File: envoy/router/router.h (hunk 2 above)
Comment (mattklein123) [reply envoy/router/router.h:386]: Yeah sorry this was causing my local push hooks to fail, not sure why this didn't fail CI. Didn't think it was worth splitting out.

File: source/common/config/filesystem_subscription_impl.cc (hunk 3 above)
Comment (mattklein123) [reply source/common/config/filesystem_subscription_impl.cc:30]: I thought about this but I'm not sure how to do this without duplicating a bunch of code that lives inside the watcher (watching a specific file, etc.). Do you have any suggestions? Happy to collapse if there is a good way. FWIW the SDS code has similar bifurcation.

File: test/config/integration/server.yaml [hunk 4]
```diff
@@ -151,7 +151,6 @@ stats_sinks:
   typed_config:
     "@type": type.googleapis.com/envoy.config.metrics.v3.StatsdSink
     tcp_cluster_name: statsd
-watchdog: {}
```
Comment (mattklein123) [inline test/config/integration/server.yaml:154]: This change and the other one like it are unrelated to this change, however when I built with disabling deprecations locally, the hot restart tests failed (we don't compile hot restart in the CI build which tests this), so I just fixed this at the same time. The empty deprecated watchdog config seems a historical legacy and not relevant to hot restart.

File: test/integration/rtds_integration_test.cc [hunk 5]
```diff
@@ -140,6 +144,70 @@ class RtdsIntegrationTest : public Grpc::DeltaSotwIntegrationParamTest, public H
...
+  TestEnvironment::createPath(temp_path + "/data_1");
+  TestEnvironment::writeStringToFileForTest(temp_path + "/data_1/rtds.yaml", R"EOF(
+resources:
+- "@type": type.googleapis.com/envoy.service.runtime.v3.Runtime
+  name: file_rtds
+  layer:
+ )EOF",
+                                            true);
+  TestEnvironment::createSymlink(temp_path + "/data_1", temp_path + "/..data");
+  TestEnvironment::createSymlink(temp_path + "/..data/rtds.yaml", temp_path + "/rtds.yaml");
```
Comment (snowp) [inline test/integration/rtds_integration_test.cc:165]: Maybe some comment with the resulting directory structure would make it easier to follow what this ends up creating?

**Review by snowp — COMMENTED** [review]
Thanks, just one small suggestion

**Review by adisuissa — COMMENTED** [review]
API LGTM, modulo comment from @htuch.

File: test/integration/rtds_integration_test.cc [hunk 6]
```diff
@@ -140,6 +144,78 @@ class RtdsIntegrationTest : public Grpc::DeltaSotwIntegrationParamTest, public H
...
+// 3. Real file `/config_map/xds/real_data/xds.yaml`
+// 4. Symlink `/config_map/xds/..data` -> `/config_map/xds/real_data`
+// 5. Symlink `/config_map/xds/xds.yaml -> `/config_map/xds/..data/xds.yaml`
+//
+// 2 symlinks are used so that multiple files can be updated in a single atomic swap of ..data to
+// a new real target.
+TEST_P(RtdsIntegrationTest, FileRtdsReload) {
+  // Create an initial setup that looks similar to a K8s ConfigMap deployment. This is a file
+  // contained in a directory and referenced via an intermediate symlink on the directory.
+  const std::string temp_path{TestEnvironment::temporaryDirectory() + "/rtds_test"};
```
Comment (htuch) [inline test/integration/rtds_integration_test.cc:163]: Nit: we have `temporaryPath()` for this.

File: source/common/config/filesystem_subscription_impl.cc (hunk 3 above)
Comment (htuch) [reply source/common/config/filesystem_subscription_impl.cc:30]: Probably not worth it for these marginal deprecated paths, so let's skip.

**Review by htuch — APPROVED** [review]
LGTM, thanks!

**Review by adisuissa — APPROVED** [review]
/lgtm api
LGTM code changes.

**Pull request merged**